### DISCUSSION TITLE
UI/UX Design and Architecture (cohices, etc)

### DISCUSSION THREAD

**Discussion by prologic** [nuxui/nuxui#6] [Q&A] [answered]
Hi 👋 It's me again!

I was wondering if you could elaborate here (firstly) on some of the design choices you're making, have already made, how teh UI/UX tookit/library works, etc.

As I watch your progress unfold and help out where I can (mostly just testing things so far) I'm curious about a couple of things:

- I see some things in sample code that looks a bit like CSS? Can you explain how this works?

The other thing I'm really curious about is how the entire library is designed in the first place, especially how the UI/UX elements work. When you look at other cross-platform systems they tend to come in a couple of varieties/flavours:

- HTML/CSS based
- OpenGL based
- Some kind of native framework itnegration, Java (Android), Objective-C/Swift/Cocoa (iOS, macOS)

How/Where does nuxui fit currently and in the future?

**Comment by sparky4pro** [comment]
Looking at the code it seems that it's a native ui framework integration, utilizing go-mobile technology and so on.

The CSS-like portion of code you mention is nothing else then a go map that describes the entities of ui components.
There are several ui libraries doing similar things (they call it declarative syntax) and I wish I could say I like it... 
Which I don't!

Using a map makes the code more convoluted the more ui components are used, everything blends-in together, which makes it harder to fix things. Same goes for the actions/events code. The correlation of ui component and their events is not 100% clear to me.

For me it's too complicated, as I like things much, much simpler...
That's why I like Go. :D

Still, tremendous work went into this already, and more to be done; hope other people will like it... and use it.

**Reply by prologic** [reply]
I have to agree with @sparkylein here actually. I'm not that big of a fan of "declarative syntax" when you're basically writing a nice set of API that integrate native UI frameworks on various platforms.

I _think_ (IMO) it would be better to pivot towards more idiomatic Go in general, even for styling.

Just my $0.02 worth 👌 I am very happy though to hear this is just wrapping native UI elements/components on various platform -- I'm _hoping_ that means we can all "features" and "experience" without having to reinvent the wheel (for UI/UX components) like in Fyne or GIO (which inevitably leads to bugs and poor experience).

**Comment by baxiry** [comment]
I think choosing a css-like ui language is very nice.  It reminds me of qml.

**Comment by nuxart** [comment, marked as answer]
## Why use css-like UI description grammar？

1. The golang struct has not default value, if use golang to write a blue background and white text button, it will be:
```go
btn := &ui.Button{
  BackgroundColor: 0x0000ffff,
  TextColor: 0xffffffff,
  Text: "clickme",
  TextSize: 16,
  Bounds: ...
}
```
It need set every filed of struct for each button when you created, maybe we can use function like:
```go
btn := NewButton("clickme")
btn.BackgroundColor = 0x0000ffff
btn.Text = 0xffffffff
```
It looks much simpler and we can use theme like:
```go
btn1 := bootstrap.NewButton("clickme")  // use theme bootstrap
btn2 := material.NewButton("clickme")  // use theme material
```
Everything looks good so far, but when we write a complete user interface, you will found your UI code, modle and controller tangled together like a spider web, you will see a bulky and complex and hard read code, so we need MVC separation design.

## Why do not use html/css directly？
1. Supported html/css need integrated webkit to nuxui, it makes nuxui bloated, and it is not golang GUI anymore, there has much better html/css GUI solution like **electron**
2. If only support html/css grammar for nuxui, it complex and easy to mislead for new developer.
3. Actually, the nuxui attr UI description grammar is **JSON**-like, but much simple than json, it just has **key** and **value** without much `"`, and can import namespace like `xmlns` of **XML**.
4. In the future, we can build a UI Editor to write UI with drag a widget like button to UI workspace without code, and auto write whole UI to a single attr file

## Why do not use openGL?
1. nuxui will support openGL in the future.
2. The windows server does not supported openGL, so it is not a urgent task.
3. The openGL is unvisible for developing, it has no effect to your current code, when nuxui supported openGL, the develop just build their app with `--tags gl` without change any code.

## Why use native framework itnegration?

1. Some fcuntions can not avoid to use native framework like open FileManager, request permissions, send notifications...
2. For the widget render, use native framework is not a good choose, it will refactor, the current code is for experimental.
3. For web widget render, maybe use native framework will take a small size of wasm. (PS. the wasm size takes more than 4MB of flutter, gio and fyne)

## What's next plan?

1. Support web.
2. Use golang render widget directly instead of native framework
3. Redesign `Widget` , becuse of golang interface{} query is so slow:
```go
if draw, ok, := widget.(Draw); ok {
    draw.Draw(canvas)
}
```
it will be
```go
if widget.Draw != nil {
  widget.Draw.Draw(canvas)
}
```
4. Complete basic widget like Row, Column, Editor....
